moveit_ros/occupancy_map_monitor/src/occupancy_map_monitor_middleware_handle.cpp

```diff
@@ -0,0 +1,138 @@
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Tyler Weaver */
+
+#include <moveit/occupancy_map_monitor/occupancy_map_monitor_middleware_handle.hpp>
+#include <pluginlib/class_loader.hpp>
+#include <string>
+#include <vector>
+
+namespace occupancy_map_monitor
+{
+namespace
+{
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit.ros.occupancy_map_monitor.middleware_handle");
+}
+
+OccupancyMapMonitorMiddlewareHandle::OccupancyMapMonitorMiddlewareHandle(const rclcpp::Node::SharedPtr& node)
+  : node_(node)
+{
+  try
+  {
+    updater_plugin_loader_ = std::make_unique<pluginlib::ClassLoader<OccupancyMapUpdater> >(
+        "moveit_ros_occupancy_map_monitor", "occupancy_map_monitor::OccupancyMapUpdater");
+  }
+  catch (pluginlib::PluginlibException& ex)
+  {
+    RCLCPP_FATAL_STREAM(LOGGER, "Exception while creating octomap updater plugin loader " << ex.what());
+    throw ex;
+  }
+}
+
+double OccupancyMapMonitorMiddlewareHandle::getMapResolutionParameter()
+{
+  double map_resolution = 0.1;
+  if (!node_->get_parameter("octomap_resolution", map_resolution))
+  {
+    RCLCPP_WARN(LOGGER, "Resolution not specified for Octomap. Assuming resolution = %g instead", map_resolution);
+  }
+  return map_resolution;
+}
+
+std::string OccupancyMapMonitorMiddlewareHandle::getMapFrameParameter()
+{
+  std::string map_frame = "";
+  node_->get_parameter("octomap_frame", map_frame);
+  return map_frame;
+}
+
+std::vector<std::string> OccupancyMapMonitorMiddlewareHandle::getSensorListParameter()
+{
+  std::vector<std::string> sensor_list;
+  if (!node_->get_parameter("sensors", sensor_list))
+  {
+    RCLCPP_ERROR(LOGGER, "No 3D sensor plugin(s) defined for octomap updates");
+  }
+  else if (sensor_list.empty())
+  {
+    RCLCPP_ERROR(LOGGER, "List of sensors is empty!");
+  }
+  return sensor_list;
+}
+
+std::string OccupancyMapMonitorMiddlewareHandle::getSensorPluginParameter(const std::string& sensor)
+{
+  std::string sensor_plugin = "";
+  if (!node_->get_parameter(sensor + ".sensor_plugin", sensor_plugin))
+  {
+    RCLCPP_ERROR(LOGGER, "No sensor plugin specified for octomap updater %s; ignoring.", sensor.c_str());
+  }
+  return sensor_plugin;
+}
+
+OccupancyMapUpdaterPtr OccupancyMapMonitorMiddlewareHandle::loadOccupancyMapUpdater(const std::string& sensor_plugin)
+{
+  OccupancyMapUpdaterPtr occupancy_map_updater = nullptr;
+  try
+  {
+    occupancy_map_updater = updater_plugin_loader_->createUniqueInstance(sensor_plugin);
```

Comment:
I generally try to avoid multiple returns if I can, but in this case, it seems simple enough and that is clearer.  Also, I would trust that the compiler would help optimize this local variable out.

Reply:
I don't usually consider what the compiler would do most of the time.
Maybe because of how I usually organize my code, multiple returns don't seem so annoying to me (I try to only return at the beginning or end). For me, the thread of extra non-const local variables is greater.

Is this why you avoid multiple returns? https://softwareengineering.stackexchange.com/questions/118703/where-did-the-notion-of-one-return-only-come-from

Reply:
It isn't something from writing assembly (I never have done much of that).  It was simply a style I found more readable.  Although in this case, I do like it better with multiple after changing it.

Reply:
Not assembly, but C

Reply:
Historical background:
The code base has/had a lot of looong if blocks just because the original authors disliked `if(!...) return;`.
I brought this up in a maintainer meeting years ago and we agreed that early return should be preferred over five layers of conditional blocks.

Reply:
Exactly. A previous codebase I worked on had a huge chain of nested blocks, which, inevitably when someone tried to maintain it, they'd get wrong (ie me). I read that part of that comes from a C style (which this codebase had originated as) where you didn't have RAII and have to worry about cleanup lest you leak resources. Nowadays I try to organize my code as
```
// Preconditions
if (condition1()) return;
if (condition2()) return;
if (condition3()) return;

// Body
do_work();

// Postconditions
if (condition4()) return 4;
return ultimate;
```
or something like that. I try to avoid returning in the body if I can.

Reply:
done